src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClients/FailOverSupportedConfigurationClient.cs

```diff
@@ -0,0 +1,186 @@
+        private readonly IEnumerable<LocalConfigurationClient> clients;
+
+        private DateTimeOffset retryPrimaryConfigClientAfter;
+
+        private int failedRequestsToPrimaryConfigClient = 0;
+
+        public FailOverSupportedConfigurationClient(IEnumerable<LocalConfigurationClient> clients)
+        {
+            if (clients == null || clients.Count() < 1)
+            {
+                throw new ArgumentNullException(nameof(clients));
+            }
+
+            this.retryPrimaryConfigClientAfter = DateTimeOffset.UtcNow;
+            this.failedRequestsToPrimaryConfigClient = 0;
+            this.clients = clients;
+        }
+
+        public Task<Response<ConfigurationSetting>> GetConfigurationSettingAsync(string key, string label = null, CancellationToken cancellationToken = default)
+        {
+            return ExecuteWithFailOverPolicyAsync(clients.Select((client) => new Func<Task<Response<ConfigurationSetting>>>(() => client.Client.GetConfigurationSettingAsync(key, label, cancellationToken))), cancellationToken);
+        }
+
+        public Task<Response<ConfigurationSetting>> GetConfigurationSettingAsync(ConfigurationSetting setting, bool onlyIfChanged = false, CancellationToken cancellationToken = default)
+        {
+            return ExecuteWithFailOverPolicyAsync(clients.Select((client) => new Func<Task<Response<ConfigurationSetting>>>(() => client.Client.GetConfigurationSettingAsync(setting, onlyIfChanged, cancellationToken))), cancellationToken);
+        }
+
+        public AsyncPageable<ConfigurationSetting> GetConfigurationSettingsAsync(SettingSelector selector, CancellationToken cancellationToken = default)
+        {
+            return ExecuteWithFailOverPolicy(clients.Select((client) => new Func<AsyncPageable<ConfigurationSetting>>(() => client.Client.GetConfigurationSettingsAsync(selector, cancellationToken))), cancellationToken);
+        }
+
+        public void UpdateSyncToken(Uri endpoint, string syncToken)
+        {
+            this.clients.Single(client => client.Endpoint.Host.Equals(endpoint.Host)).Client.UpdateSyncToken(syncToken);
+        }
+
+        private async Task<Response<T>> ExecuteWithFailOverPolicyAsync<T>(IEnumerable<Func<Task<Response<T>>>> delegates, CancellationToken cancellationToken = default)
+        {
+            Exception latestException = null;
+
+            for (var i = ShouldTryPrimaryConfigStore() ? 0 : 1; i < clients.Count(); i++)
+            {
+                var success = false;
+                var operationCanceled = false;
+
+                try
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+                    var result = await delegates.ElementAt(i)().ConfigureAwait(false);
+                    success = true;
+
+                    return result;
+                }
+                catch (RequestFailedException e) when (e.Status == (int)HttpStatusCode.ServiceUnavailable || e.Status == HttpStatusRequestThrottled)
```

Comment:
Should we try other replicas for any >500 errors, not just 503 (ServiceUnavailable)?

Reply:
With the latest design, we retry on other errors as well.